### PR TITLE
[Feature] - Platform V4 Markdown Generator V2

### PR DESCRIPTION
**Pull request by the-praxs**
# [Feature] - Platform V4 Markdown Generator V2
1. **Why**?

    - To improve the markdown generation by removing the use of docstrings of the static files
    - To remove dependency on various classes
    - To improve the code structure and readability

2. **What**?

    - Re-built the markdown generator from the ground up mainly using the `ProviderInterface` and `APIRouter` classes
    - Reduced imports of several packages and classes thereby reducing dependencies
    - Improved the code structure by creating functions for specific use cases
    - Improved code readability by using clear variable and function names, adding comments and docstrings wherever desired.

3. **Impact**

    - High Impact
    
    - Impact Analysis breakdown:
        - _UX_ = 0 (Developer use only)
        - _Performance_ = 3 (Reduced package imports)
        - _Compatibility_ = 5 (Re-built from scratch)
        - _Security_ = 3 (File System Operations, Extra JSON generation)
        - _Scalability_ = 4 (Scalable to software leveraging the Platform Python interface)
        - _Maintenance_ = 2 (Maintainance shifted towards the Platform core)
        - _Compliance_ = 0 (Compliant with the current license)
        - _Dependency_ = 5 (Reduced dependency on external packages and several internal classes)

    - _Total Impact Score_ = 22

4. **Testing Done**

    - Static built and validated for all the extensions and providers
    - Functions tested individually manually
    - Script tested for performance and edge cases manually
    - Generated files tested extensively by locally building the documentation files via `npm`

5. **Reviewer Notes**

    Testing procedure - 
    - Ensure you are on the latest version of the `develop` branch
    - Use the following command to install all the extensions and providers locally -
      ````bash
      python openbb_platform/dev_install.py -e
      ````

    - Go to the `website` directory and run the following command to generate the platform markdown files -
      ```bash
      python generate_platform_v4_markdown.py
      ```

    - Build the documentation for the generated packages using the following command -
      ```bash
      npm run dev
      ```

### CODE REVIEW

File: openbb_platform/providers/yfinance/openbb_yfinance/models/company_news.py [hunk 1]
```diff
@@ -33,7 +33,7 @@ class YFinanceCompanyNewsData(CompanyNewsData):
     uuid: str = Field(description="Unique identifier for the news article")
     publisher: str = Field(description="Publisher of the news article")
     type: str = Field(description="Type of the news article")
-    thumbnail: Optional[List] = Field(
+    thumbnail: Optional[List[Dict[str, Any]]] = Field(
         default=None, description="Thumbnail related data to the ticker news article."
```
Comment (deeleeramone) [inline openbb_platform/providers/yfinance/openbb_yfinance/models/company_news.py:37]: This field needs standardization between all providers. We have:

- "image"
- "image_url"
- "thumbnail"
- "images"

Comment (IgorWounds) [reply openbb_platform/providers/yfinance/openbb_yfinance/models/company_news.py:37]: @hjoaquim has a ticket that deals with this. @deeleeramone maybe also open an issue on our bugs ticket.

File: website/generate_platform_v4_markdown.py
Comment (IgorWounds) [inline website/generate_platform_v4_markdown.py:1]: Be careful how you handle paths. "/" won't work well on Windows.

**Comment by IgorWounds** [comment]
@the-praxs please ensure the pathing supports Windows too. Also, it is advisable to add tests for this.

File: openbb_platform/providers/yfinance/openbb_yfinance/models/company_news.py (hunk 1 above)
Comment (hjoaquim) [reply openbb_platform/providers/yfinance/openbb_yfinance/models/company_news.py:37]: Agreed. But please, let's open a new issue and keep this PR within its scope!!

File: website/generate_platform_v4_markdown.py
Comment (the-praxs) [reply website/generate_platform_v4_markdown.py:1]: The `path` package allows us interoperability. I have tested this script on Windows and have found no issues so far.

**Review by montezdesousa — CHANGES_REQUESTED** [review]
As we spoke offline i'd move the `reference.json` generation to package builder and ensure consistency for standard parameters like dates being `Union[str, date]` instead of just `date`.

Besides that the code looks much cleaner, nice job

**Comment by the-praxs** [comment]
> Overall it's looking good, and the script runs really fast. I have found that the card titles are cut off where the title is long. The underscore would make word wrap look weird, so not sure what the solution is. Is there an "autofit" for font size and content?
> 
> 
> 
> Screenshot 2024-02-21 at 1 04 22 PM
> 
> 

Thanks for reviewing the script!

My rationale for the underscore was to make it look similar like the commands. I am unsure about the "autofit" for font size and content at the moment but if the problem persists I can easily remove the underscores.

**Comment by deeleeramone** [comment]
> My rationale for the underscore was to make it look similar to the commands. I am unsure about the "autofit" for font size and content at the moment but if the problem persists I can easily remove the underscores.

I agree with the rationale, but maybe the best course of action is to remove the underscores so that the title text can wrap.  I'd want to keep the sidebar titles though.

**Comment by the-praxs** [comment]
> > My rationale for the underscore was to make it look similar to the commands. I am unsure about the "autofit" for font size and content at the moment but if the problem persists I can easily remove the underscores.
> 
> I agree with the rationale, but maybe the best course of action is to remove the underscores so that the title text can wrap. I'd want to keep the sidebar titles though.

Done.

image

**Review by deeleeramone — CHANGES_REQUESTED** [review]
I'm getting an error on build that appears to be single quotes interfering with the description text. Params should all be wrapped with " instead.

Screenshot 2024-02-22 at 12 51 19 PM

**Comment by the-praxs** [comment]
> I'm getting an error on build that appears to be single quotes interfering with the description text. Params should all be wrapped with " instead.
> 
> Screenshot 2024-02-22 at 12 51 19 PM

Wrapped them with `"` but the description for `econometrics/causality` is failing the build. Fixing it right now!
image

**Comment by deeleeramone** [comment]
> Wrapped them with `"` but the description for `econometrics/causality` is now failing the build. Fixing it right now! !=

Ya, unfortunately Replace All will render the single quotes that we want to keep as double. Can't win 'em all!

**Comment by the-praxs** [comment]
> > Wrapped them with `"` but the description for `econometrics/causality` is now failing the build. Fixing it right now! !=
> 
> Ya, unfortunately Replace All will render the single quotes that we want to keep as double. Can't win 'em all!

Gotta catch 'em all! Looks good now.

**Review by deeleeramone — APPROVED** [review]
This looks good to me, I've only tested on a Mac though. If we can sort the items under the Reference sidebar, I'll never have to struggle to find where an item is.

**Comment by the-praxs** [comment]
> This looks good to me, I've only tested on a Mac though. If we can sort the items under the Reference sidebar, I'll never have to struggle to find where an item is.

I tested on Windows and it's working as expected.

The Reference sidebar is sorted alphabetically on the main and nested levels.

image

**Review by montezdesousa — APPROVED** [review]

**Comment by jmaslek** [comment]
On the side panel, I end up with what looks like docstrings instead of the models.

Screenshot 2024-02-23 at 10 56 46 AM

**Comment by jmaslek** [comment]
> Both these columns are explicit about the `Data` fields.
> 
> _Optional_ conveys that you may or may not expect the field to be returned for a particular provider. In case a standard model field is overridden in the provider model and it is supposedly non-optional the user must know about it.
> 
> _Default_ I can see having not much use in the current scenario but maybe in the future for a data model.

I agree with Andrew here,  not sure what that gives for the `Data` models, especially when everything seems to end up as optional with no default.

**Comment by the-praxs** [comment]
> > Both these columns are explicit about the `Data` fields.
> > _Optional_ conveys that you may or may not expect the field to be returned for a particular provider. In case a standard model field is overridden in the provider model and it is supposedly non-optional the user must know about it.
> > _Default_ I can see having not much use in the current scenario but maybe in the future for a data model.
> 
> I agree with Andrew here, not sure what that gives for the `Data` models, especially when everything seems to end up as optional with no default.

Removing it for the time being then.

**Comment by the-praxs** [comment]
> On the side panel, I end up with what looks like docstrings instead of the models.
> 
> Screenshot 2024-02-23 at 10 56 46 AM

Can you tell how to reproduce this?

**Comment by jmaslek** [comment]
> > On the side panel, I end up with what looks like docstrings instead of the models.
> > Screenshot 2024-02-23 at 10 56 46 AM
> 
> Can you tell how to reproduce this?

launch the docs locally

**Comment by the-praxs** [comment]
> > > On the side panel, I end up with what looks like docstrings instead of the models.
> > > Screenshot 2024-02-23 at 10 56 46 AM
> > 
> > 
> > Can you tell how to reproduce this?
> 
> launch the docs locally

Those are the files from a very very early Platform version. Please follow the instructions under the **Reviewer Notes** in the PR description.

**Review by andrewkenreich — APPROVED** [review]

**Comment by the-praxs** [comment]
> > > This looks good to me, I've only tested on a Mac though. If we can sort the items under the Reference sidebar, I'll never have to struggle to find where an item is.
> > 
> > 
> > I tested on Windows and it's working as expected.
> > The Reference sidebar is sorted alphabetically on the main and nested levels.
> 
> Does not seem to appear that way for me on Mac.
> 
> Screenshot 2024-02-23 at 8 37 41 AM

Let me try to reproduce on my Mac.

**Comment by hjoaquim** [comment]
> > > > On the side panel, I end up with what looks like docstrings instead of the models.
> > > > Screenshot 2024-02-23 at 10 56 46 AM
> > > 
> > > 
> > > Can you tell how to reproduce this?
> > 
> > 
> > launch the docs locally
> 
> Those are the files from a very very early Platform version. Please follow the instructions under the **Reviewer Notes** in the PR description.

maybe we should remove them?

**Comment by the-praxs** [comment]
> > > > > On the side panel, I end up with what looks like docstrings instead of the models.
> > > > > Screenshot 2024-02-23 at 10 56 46 AM
> > > > 
> > > > 
> > > > Can you tell how to reproduce this?
> > > 
> > > 
> > > launch the docs locally
> > 
> > 
> > Those are the files from a very very early Platform version. Please follow the instructions under the **Reviewer Notes** in the PR description.
> 
> maybe we should remove them?

@IgorWounds thoughts?

**Pull request merged**